### ISSUE TITLE
Allow to remove Failed pods with no OwnerRef

### ISSUE THREAD

**Issue by weisdd** [kubernetes-sigs/descheduler#644]
**Is your feature request related to a problem? Please describe.**
Currently, Failed pods with no OwnerRef cannot be evicted by RemoveFailedPods policy.
Manually created pods are taken out of consideration as they will not be recreated by controllers. Though, I think it should not be a concern for Failed pods.

**Describe the solution you'd like**
Make Failed pods evictable despite lack of OwnerRef.

**Describe alternatives you've considered**
Manually deleting pods

**What version of descheduler are you using?**
descheduler version: 0.22.1

**Comment by damemi** [comment]
I [commented in your PR](https://github.com/kubernetes-sigs/descheduler/pull/645#pullrequestreview-775000570) before I saw this issue, but just for reference here applying the [`descheduler.alpha.kubernetes.io/evict` annotation](https://github.com/kubernetes-sigs/descheduler#pod-evictions) to any pod will make it evictable regardless of the various eviction checks. So I don't think this is necessary for just one strategy

**Comment by weisdd** [comment]
@damemi well, when there are many people experimenting in the same dev-cluster, it's hard to force everyone to always add the annotation. E.g. when people create Jobs and then never clean them up. So, I thought it would be much easier to simply deschedule Failed pods in that case. Setting up mutation webhook would be way more difficult.

**Comment by damemi** [comment]
So we don't have to duplicate the comments, I propose we keep the discussion for this in this issue rather than the PR 🙂  From there (so others can see):

>@weisdd I understand your point, but couldn't the same be said for other strategies like PodLifetime? Maybe we should delineate a group of strategies that apply to "expired" pods to consider this kind of behavior.
>
>If we do add this, it should be as a non-default option since it is a change in behavior from the current default.

**Comment by weisdd** [comment]
> So we don't have to duplicate the comments, I propose we keep the discussion for this in this issue rather than the PR

@damemi agreed, was about to suggest the same :)

I've actually used descheduler for the first time today (it's been on my list of things to try for ages), so my perception of how things can potentially be done might be too far from the project's maintainers vision :)

Making Failed pods evictable felt in line with what I'd expect from descheduler. The change in the PR seemed to be the easiest solution and, which is more important, to be based on a safe assumption that nobody would want to have those pods around for longer than `minPodLifeTimeSeconds` (+ safe enough to make it default). Though, I'm open to whatever feels better to you.

Again, since I'm not very familiar with the code base, not sure how to properly design that delineation of strategies. :sweat_smile: 

As for making the option non-default, maybe introduction of something like: `evictFailedPodsWithNoOwnerRef: false` (as a policy) would work.

**Comment by damemi** [comment]
Yeah I think there's definitely a useful solution here. It's an interesting suggestion, so I'd like to see if we can get some input from others on possibly fleshing it out a bit more. @ingvagabund @seanmalloy wdyt?

**Comment by ingvagabund** [comment]
> Make Failed pods evictable despite lack of OwnerRef.

Reading this sentence again, it makes sense to allow to remove these pods as well. Though, only as an option. Such pods still have logs which can be used for debugging.

/good-first-issue

**Comment by babygoat** [comment]
Hi @ingvagabund, @weisdd 
How about we add a flag `RemoveBarePod`  and default to `false` when initializing the `DeshedulerServer`? Then, define the `removeBarePod` flag on `PodEvictor` to accept the configuration from the command. Finally, we move the[original constraint](https://github.com/kubernetes-sigs/descheduler/blob/e0bac4c37182e683c26fa16446779695517e140a/pkg/descheduler/evictions/evictions.go#L281-283)
to Evictable() to add constraints depending on the `removeBarePod`.
```go
...
ownerRefList := podutil.OwnerRef(pod)
if !pe.removeBarePod {
   	ev.constraints = append(ev.constraints, func(pod *v1.Pod) error {
		if  len(ownerRefList) == 0 {
			return fmt.Errorf("pod does not have any ownerrefs")
		}
		return nil
	})
}
...
```
I was wondering if these are the direction we would like to move forward. If that is OK, I could help with the change.

**Comment by damemi** [comment]
@babygoat I think that general approach sounds good, though rather than a flag this should be a Policy field (we're moving away from config flags)

@ingvagabund wdyt?

**Comment by ingvagabund** [comment]
Definitely a Policy field. The approach looks good.

**Comment by babygoat** [comment]
Thank @ingvagabund and @damemi  for your suggestions!  I will provide the PR later to address with the Policy approach.

**Comment by weisdd** [comment]
Was about to say that I should be able to work on this soon, though thanks @babygoat for picking this up, I'll wait for your PR then :)

**Comment by babygoat** [comment]
Hi @weisdd, thanks for your willingness to let me try. I was wondering if we could use the #688 to address the issue instead of your original PR #645.

**Comment by seanmalloy** [comment]
Implemented by #688

/close
